### ISSUE TITLE
Alpaca (Fine-tuned LLaMA)

### ISSUE THREAD

**Issue by slavakurilyak** [langchain-ai/langchain#1777]
It would be great to see LangChain integrate with Standford's Alpaca 7B model, a fine-tuned LlaMa (see https://github.com/hwchase17/langchain/issues/1473).

Standford created an AI able to generate outputs that were largely on par with OpenAI’s `text-davinci-003` and regularly better than GPT-3 — all for a fraction of the computing power and price.

Stanford's Alpaca is a language model that was fine-tuned from Meta's LLaMA with 52,000 instructions generated by GPT-3.5[[1]](https://the-decoder.com/stanfords-alpaca-shows-that-openai-may-have-a-problem/). The researchers used AI-generated instructions to train Alpaca 7B, which exhibits many GPT-3.5-like behaviors[[1]](https://the-decoder.com/stanfords-alpaca-shows-that-openai-may-have-a-problem/). In a blind test using input from the Self-Instruct Evaluation Set, both models performed comparably[[1]](https://the-decoder.com/stanfords-alpaca-shows-that-openai-may-have-a-problem/)[[2]](https://github.com/tatsu-lab/stanford_alpaca). However, Alpaca has problems common to other language models such as hallucinations, toxicity, and stereotyping[[1]](https://the-decoder.com/stanfords-alpaca-shows-that-openai-may-have-a-problem/). The team is releasing an interactive demo, the training dataset, and the training code for research purposes[[1]](https://the-decoder.com/stanfords-alpaca-shows-that-openai-may-have-a-problem/)[[4]](https://crfm.stanford.edu/2023/03/13/alpaca.html).

Alpaca is still under development and has limitations that need to be addressed. The researchers have not yet fine-tuned the model to be safe and harmless[[2]](https://github.com/tatsu-lab/stanford_alpaca). They encourage users to be cautious when interacting with Alpaca and report any concerning behavior to help improve it[[2]](https://github.com/tatsu-lab/stanford_alpaca).

LLaMA is a new open-source language model from Meta Research that performs as well as closed-source models. Stanford's Alpaca is a fine-tuned version of LLaMA that can respond to instructions like ChatGPT[[3]](https://replicate.com/blog/replicate-alpaca). It functions more like a fancy version of autocomplete than a conversational bot[[3]](https://replicate.com/blog/replicate-alpaca).

The researchers are releasing their findings about an instruction-following language model dubbed Alpaca. They trained the Alpaca model on 52K instruction-following demonstrations generated in the style of self-instruct using `text-davinci-003`. On the self-instruct evaluation set, Alpaca shows many behaviors similar to OpenAI’s `text-davinci-003` but is also surprisingly small and easy/cheap to reproduce[[4]](https://crfm.stanford.edu/2023/03/13/alpaca.html).

**Comment by lorepieri8** [comment]
Also worth considering Instruct-tuned LLaMA models that can run on consumer hardware, such as 
https://github.com/tloen/alpaca-lora.

**Comment by linonetwo** [comment]
A basic LangChainJS package for Alpaca is here https://github.com/linonetwo/langchain-alpaca

**Comment by iRanadheer** [comment]
> A basic LangChainJS package for Alpaca is here https://github.com/linonetwo/langchain-alpaca

Great!! but would be nicer if there's a python wrapper

**Comment by lukestanley** [comment]
I started on hacking a Python version based on Serge, which is MIT licensed. Consider my hack MIT licensed also.
https://gist.github.com/lukestanley/6517823485f88a40a09979c1a19561ce

**Comment by 1b5d** [comment]
I've written an app to run llama based models using docker here: https://github.com/1b5d/llm-api 
You can specify the model in the config file, and the app will download it automatically and expose it via an API
Additionally you can use https://github.com/1b5d/langchain-llm-api in order to use this exposed API with Langchain, it also supports streaming
My goal is to easily run different models locally (and also remote) and switch between them easily, then use these APIs to develop with Langchain

**Comment by lukestanley** [comment]
I realised there is a Llamma class already integrated. Better put something in my eyes so I don't need to blink!
I have a draft PR that adds streaming here: https://github.com/hwchase17/langchain/pull/2775

**Comment by annabechang** [comment]
there's alpaca on huggingface, i think you can do something like this: 
`llm = HuggingFaceHub(repo_id="chavinlo/alpaca-native", model_kwargs={"temperature":0, "max_length":64})`

**Comment by iRanadheer** [comment]
you can use the huggingfacepipeline class.

**Comment by tensiondriven** [comment]
:clap: Thank you @iRanadheer  - Information on how to integrate local models with langchain is sparse.  This little tidbit is helpful.  If you know of any repos that provide toy examples of such integration, please post a link!

**Comment by lukestanley** [comment]
The LangChain repo has documentation on how to use LlamaCpp within LangChain with a LLM class like the OpenAI one.
The LlamaCpp LLM class is probably faster and easier to setup for most people, and is documented here:
https://python.langchain.com/en/latest/modules/models/llms/integrations/llamacpp.html
Like the underlying code, it can generate with Alpaca models too.
I am working on adding streaming support to it [here](https://github.com/hwchase17/langchain/pull/2775).
This is happening fast and it might be hard to find, I think this issue might deserve closing soon since it works.
I'll probably make a gif of it working with a funny example.
Any ideas on other ways it could be more visible?
@tensiondriven

**Comment by tensiondriven** [comment]
@lukestanley Thanks for asking!

I saw the LlamaCpp example, but my understanding is that the -Cpp interface/classes are very different from huggingface/pt/safetensors.

I agree that Cpp is more accessible for more people, but it doesn't address what I think is a pretty common use case: Using HF-format models locally for GPU inference with CUDA/Triton.

When I read the docs, I saw the bit about "integrating with runway, and also local" (was it called runway?  I'm not sure, I'm on my phone at the moment)

What would have made it more clear is if there was a specific page for "Local LLM" that gives an example which takes a path to a safetensors file and includes setup of the required dependencies (pytorch/fromCasusalLangaugModel, etc.)

Bonus points if it includes an example for loading with GPTQ 4-bit, as a lot of people need to run 4-bit locally due to limited VRAM.

Extra bonus points if it includes an example of loading a safetensors model with a Lora.

I realize these implementation details are not specifically a concern of langchain, however given how fiddly the python deps are, I still think it would be useful to show a full example of standing up a GPU-based local model using a safetensors model as its own page in the docs under LLM's.

As a side-note, there are a few bugs in the chat-demo project, which there are fixes for in the GitHub Issues, but those small bugs do add significant friction for people who want to get langchain up and running (with either OpenAI or a different/local LLM).

For one additional bit of context, I think a lot of people are coming from using text-generation-webui, which, while being pretty poorly organized code, at least gives people exposure to the python dependency stack and some of the parameters used in loading language models.  For me, getting these dependencies and parameters right was difficult, and the langchain docs brush over this.  I found the above comment very useful since it showed how to load an LLM in python and then pass it to a Langchain LLM adapter (I know it's trivial to do, I still found it to be the missing link for me)

For a little more context on where I'm coming from; 10-year ruby/JavaScript/elixir dev, professional and hobbiest, with some python experience, but python is not my first language.

Tldr; Need a Local HF safetensors/pt example with all the necessary imports alongside all the hosted LLM examples

Does this help?

(Also, very excited about streaming support for local LLM's! Will be happy to test this if that's helpful)

**Comment by bowenwen** [comment]
For those of you who are interested in using transformer with langchain's HuggingFacePipeline, I was able to run Llama locally using the HuggingFacePipeline provided as part of langchain.llms. It works pretty well using the 8-bit models, and I have an example use here:
https://github.com/bowenwen/llama-at-home/blob/4d550a738136b7e9a23dc1f2b9bbf95367f504e2/src/my_langchain_models.py#L135-L214

If you would like to also use lora (with alpaca fine tuning), it is possible but you will get a warning message saying PeftModelForCausalLM is not supported for the transformer pipeline, but as far as I can tell, everything works fine. This is how I implemented Peft model with lora:
https://github.com/bowenwen/llama-at-home/blob/4d550a738136b7e9a23dc1f2b9bbf95367f504e2/src/my_langchain_models.py#L65-L99

I also have some wrappers I wrote to help with integrating various components of langchain like docs with vectorstore and agent executors. Feel free to browse around my repo and grab anything you'd like.
